doc/api/module.md

```diff
@@ -66,6 +66,46 @@ const require = createRequire(import.meta.url);
 const siblingModule = require('./sibling-module');
 ```
 
+### `module.detectSyntax(code)`
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.0 - Early development
+
+The `module.detectSyntax()` method attempts to determine the syntax of agiven source code string.
+This detection is performed without executing the code, relying on file extensions, or checking the type field in `package.json`.
+Only TypeScript syntax that can be erased is supported.
+
+* `code` {string} The code to detect the syntax of.
+* Returns: {string|undefined} The detected syntax of the code. Possible values are:
+  * `'commonjs'` for CommonJS module code.
+  * `'module'` for ECMAScript module code.
+  * `'commonjs-typescript'` for CommonJS TypeScript code.
+  * `'module-typescript'` for ECMAScript TypeScript code.
+  * `undefined` if the syntax cannot be detected.
+    **Default:** `'commonjs'`.
+
+```mjs
+import { detectSyntax } from 'node:module';
+detectSyntax('export const a = 1;'); // 'module'
+detectSyntax('const a = 1;'); // 'commonjs'
+detectSyntax('export const a: number = 1;'); // 'module-typescript'
+detectSyntax('const a: number = 1;'); // 'commonjs-typescript'
+detectSyntax('const foo;'); // Invalid syntax, returns undefined
+```
+
+There are some limitations to the detection, where the syntax cannot be determined because it's ambiguous:
+
+```mjs
+import { detectSyntax } from 'node:module';
+// This could be CommonJS if we interpret < and > as greater and lesser,
+// or it could be TypeScript if we interpret them as type parameters.
+// In this case, it will be interpreted as TypeScript.
```

Comment:
this seems potentially problematic, since TS isn't a syntactic superset of JS. Perhaps in this case it could return an array of two choices?

Reply:
it's an edge case, how do you detect it though 😄 I guess it has to be clear its heuristic, it trades some correctness for convenience

Reply:
if it can be parsed as both, then it's the edge case.

since this function is intended to be used by libraries and not by most users, it seems like correctness is the better tradeoff to make here. Doing `result?.[0] ?? result` isn't that inconvenient, but allows the choice and visibility into the edge cases.

Reply:
the only way I see to disambiguate the syntax is to `compileScript` and see if it works, and attempt to strip types and if the output is different, it means its an edge case.

Reply:
Sure, that seems fine to me.